pkg/kubectl/cmd/set/set_data.go

```diff
@@ -0,0 +1,291 @@
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+// Run creates and applies the patch either Locally or calling apiserver.
+func (o *DataOptions) Run() error {
+	patches := CalculatePatches(o.Infos, cmdutil.InternalVersionJSONEncoder(), func(info *resource.Info) ([]byte, error) {
+		transformed, err := o.updateDataForObject(info.Object, addData)
+		if transformed && err == nil {
+			return runtime.Encode(cmdutil.InternalVersionJSONEncoder(), info.AsVersioned())
+		}
+		return nil, err
+	})
+
+	allErrs := []error{}
+	for _, patch := range patches {
+		info := patch.Info
+		if patch.Err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error: %s/%s %v\n", info.Mapping.Resource, info.Name, patch.Err))
+			continue
+		}
+
+		//no changes
+		if string(patch.Patch) == "{}" || len(patch.Patch) == 0 {
+			allErrs = append(allErrs, fmt.Errorf("info: %s %q was not changed\n", info.Mapping.Resource, info.Name))
+			continue
+		}
+
+		if o.Local || o.DryRun {
+			if err := cmdutil.PrintObject(o.Cmd, patch.Info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+
+		obj, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, types.StrategicMergePatchType, patch.Patch)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("failed to patch data to : %v\n", err))
+			continue
+		}
+		info.Refresh(obj, true)
+
+		if len(o.Output) > 0 {
+			if err := cmdutil.PrintObject(o.Cmd, info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+		cmdutil.PrintSuccess(o.ShortOutput, o.Out, info.Object, o.DryRun, "data updated")
+	}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+func getData(objType string, obj interface{}, fileSources, literalSources []string, envFileSource string) error {
+	if objType == "secret" {
+		if len(fileSources) > 0 {
+			if err := kubectl.HandleFromFileSources(obj.(*v1.Secret), fileSources); err != nil {
+				return err
+			}
+		}
+		if len(literalSources) > 0 {
+			if err := kubectl.HandleFromLiteralSources(obj.(*v1.Secret), literalSources); err != nil {
+				return err
+			}
+		}
+		if len(envFileSource) > 0 {
+			if err := kubectl.HandleFromEnvFileSource(obj.(*v1.Secret), envFileSource); err != nil {
+				return err
+			}
+		}
+	} else if objType == "configmap" {
+		if len(fileSources) > 0 {
+			if err := kubectl.HandleConfigMapFromFileSources(obj.(*v1.ConfigMap), fileSources); err != nil {
+				return err
+			}
+		}
+		if len(literalSources) > 0 {
+			if err := kubectl.HandleConfigMapFromLiteralSources(obj.(*v1.ConfigMap), literalSources); err != nil {
+				return err
+			}
+		}
+		if len(envFileSource) > 0 {
+			if err := kubectl.HandleConfigMapFromEnvFileSource(obj.(*v1.ConfigMap), envFileSource); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (o *DataOptions) updateDataForObject(obj runtime.Object, fn func(string, runtime.Object, interface{}) (bool, map[string]string, map[string][]byte)) (bool, error) {
```

Comment:
The problem with this function is that it's doing two very distinct things (one thing for Secret, one thing for Configmap). It should be two different functions instead. Also, this function is called only once, I still don't understand the point of taking a function as a parameter.